engines/sherlock/screen.cpp

```diff
@@ -0,0 +1,534 @@
+}
+
+/**
+ * Fades from the currently active palette to the passed palette
+ */
+int Screen::equalizePalette(const byte palette[PALETTE_SIZE]) {
+	int total = 0;
+	byte tempPalette[PALETTE_SIZE];
+	getPalette(tempPalette);
+
+	// For any palette component that doesn't already match the given destination
+	// palette, change by 1 towards the reference palette component
+	for (int idx = 0; idx < PALETTE_SIZE; ++idx) {
+		if (tempPalette[idx] > palette[idx])
+		{
+			tempPalette[idx] = MAX((int)palette[idx], (int)tempPalette[idx] - 4);
+			++total;
+		} else if (tempPalette[idx] < palette[idx]) {
+			tempPalette[idx] = MIN((int)palette[idx], (int)tempPalette[idx] + 4);
+			++total;
+		}
+	}
+
+	if (total > 0)
+		// Palette changed, so reload it
+		setPalette(tempPalette);
+
+	return total;
+}
+
+/**
+ * Fade out the palette to black
+ */
+void Screen::fadeToBlack(int speed) {
+	byte tempPalette[PALETTE_SIZE];
+	Common::fill(&tempPalette[0], &tempPalette[PALETTE_SIZE], 0);
+
+	while (equalizePalette(tempPalette)) {
+		_vm->_events->delay(15 * speed);
+	}
+
+	setPalette(tempPalette);
+	fillRect(Common::Rect(0, 0, this->w, this->h), 0);
+}
+
+/**
+ * Fade in a given palette
+ */
+void Screen::fadeIn(const byte palette[PALETTE_SIZE], int speed) {
+	int count = 50;
+	while (equalizePalette(palette) && --count) {
+		_vm->_events->delay(15 * speed);
+	}
+
+	setPalette(palette);
+}
+
+/**
+ * Adds a rectangle to the list of modified areas of the screen during the
+ * current frame
+ */
+void Screen::addDirtyRect(const Common::Rect &r) {
+	_dirtyRects.push_back(r);
+	assert(r.width() > 0 && r.height() > 0);
+}
+
+/**
+ * Merges together overlapping dirty areas of the screen
+ */
+void Screen::mergeDirtyRects() {
+	Common::List<Common::Rect>::iterator rOuter, rInner;
+
+	// Process the dirty rect list to find any rects to merge
+	for (rOuter = _dirtyRects.begin(); rOuter != _dirtyRects.end(); ++rOuter) {
+		rInner = rOuter;
+		while (++rInner != _dirtyRects.end()) {
+
+			if ((*rOuter).intersects(*rInner)) {
+				// these two rectangles overlap or
+				// are next to each other - merge them
+
+				unionRectangle(*rOuter, *rOuter, *rInner);
+
+				// remove the inner rect from the list
+				_dirtyRects.erase(rInner);
+
+				// move back to beginning of list
+				rInner = rOuter;
+			}
+		}
+	}
+}
+
+/**
+ * Returns the union of two dirty area rectangles
+ */
+bool Screen::unionRectangle(Common::Rect &destRect, const Common::Rect &src1, const Common::Rect &src2) {
+	destRect = src1;
+	destRect.extend(src2);
+
+	return !destRect.isEmpty();
+}
+
+/**
+ * Do a random pixel transition in from _backBuffer surface to the screen
+ */
+void Screen::randomTransition() {
+	Events &events = *_vm->_events;
+	const int TRANSITION_MULTIPLIER = 0x15a4e35;
+	_dirtyRects.clear();
+
+	for (int idx = 0; idx <= 65535 && !_vm->shouldQuit(); ++idx) {
+		_transitionSeed = _transitionSeed * TRANSITION_MULTIPLIER + 1;
+		int offset = _transitionSeed & 65535;
```

Comment:
Personally, I would suggest using a hex number here so it's easier to see which bits are affected.